File: libs/connecteddevice/src/com/google/android/connecteddevice/storage/ConnectedDeviceStorage.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.connecteddevice.storage;

import static com.google.android.connecteddevice.util.SafeLog.logd;
import static com.google.android.connecteddevice.util.SafeLog.loge;
import static com.google.android.connecteddevice.util.SafeLog.logw;

import android.app.ActivityManager;
import androidx.room.Room;
import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import com.google.android.connecteddevice.model.AssociatedDevice;
import java.security.InvalidKeyException;
import java.security.InvalidParameterException;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/** Storage for connected devices in a car. */
public class ConnectedDeviceStorage {
  private static final String TAG = "CompanionStorage";

  private static final String SHARED_PREFS_NAME = "com.google.android.connecteddevice";
  private static final String UNIQUE_ID_KEY = "CTABM_unique_id";
  private static final String DATABASE_NAME = "connected-device-database";

  private static final String CHALLENGE_HASHING_ALGORITHM = "HmacSHA256";

  @VisibleForTesting public static final int CHALLENGE_SECRET_BYTES = 32;

  private final Context context;

  private final AssociatedDeviceDao associatedDeviceDatabase;

  private final CryptoHelper cryptoHelper;

  private SharedPreferences sharedPreferences;

  private UUID uniqueId;

  private AssociatedDeviceCallback associatedDeviceCallback;

  public ConnectedDeviceStorage(@NonNull Context context) {
    this(
        context,
        new KeyStoreCryptoHelper(),
        Room.databaseBuilder(context, ConnectedDeviceDatabase.class, DATABASE_NAME)
            .fallbackToDestructiveMigration()
            .build()
            .associatedDeviceDao());
  }

  @VisibleForTesting
  public ConnectedDeviceStorage(
      @NonNull Context context,
      @NonNull CryptoHelper cryptoHelper,
      @NonNull AssociatedDeviceDao associatedDeviceDatabase) {
    this.context = context;
    this.cryptoHelper = cryptoHelper;
    this.associatedDeviceDatabase = associatedDeviceDatabase;
  }

  /**
   * Set a callback for associated device updates.
   *
   * @param callback {@link AssociatedDeviceCallback} to set.
   */
  public void setAssociatedDeviceCallback(@NonNull AssociatedDeviceCallback callback) {
    associatedDeviceCallback = callback;
  }

  /** Clear the callback for association device callback updates. */
  public void clearAssociationDeviceCallback() {
    associatedDeviceCallback = null;
  }

  /**
   * Get communication encryption key for the given device.
   *
   * @param deviceId id of trusted device
   * @return encryption key, null if device id is not recognized
   */
  @Nullable
  public byte[] getEncryptionKey(@NonNull String deviceId) {
    AssociatedDeviceKeyEntity entity = associatedDeviceDatabase.getAssociatedDeviceKey(deviceId);
    if (entity == null) {
      logd(TAG, "Encryption key not found!");
      return null;
    }

    return cryptoHelper.decrypt(entity.encryptedKey);
  }

  /**
   * Save encryption key for the given device.
   *
   * @param deviceId id of the device
   * @param encryptionKey encryption key
   */
  public void saveEncryptionKey(@NonNull String deviceId, @NonNull byte[] encryptionKey) {
    String encryptedKey = cryptoHelper.encrypt(encryptionKey);
    AssociatedDeviceKeyEntity entity = new AssociatedDeviceKeyEntity(deviceId, encryptedKey);
    associatedDeviceDatabase.addOrReplaceAssociatedDeviceKey(entity);
    logd(TAG, "Successfully wrote encryption key.");
  }

  /**
   * Save challenge secret for the given device.
   *
   * @param deviceId id of the device
   * @param secret Secret associated with this device. Note: must be {@value CHALLENGE_SECRET_BYTES}
   *     bytes in length or an {@link InvalidParameterException} will be thrown.
   */
  public void saveChallengeSecret(@NonNull String deviceId, @NonNull byte[] secret) {
    if (secret.length != CHALLENGE_SECRET_BYTES) {
      throw new InvalidParameterException(
          "Secrets must be " + CHALLENGE_SECRET_BYTES + " bytes in length.");
    }

    String encryptedKey = cryptoHelper.encrypt(secret);
    AssociatedDeviceChallengeSecretEntity entity =
        new AssociatedDeviceChallengeSecretEntity(deviceId, encryptedKey);
    associatedDeviceDatabase.addOrReplaceAssociatedDeviceChallengeSecret(entity);
    logd(TAG, "Successfully wrote challenge secret.");
  }

  /** Get the challenge secret associated with a device. */
  public byte[] getChallengeSecret(@NonNull String deviceId) {
    AssociatedDeviceChallengeSecretEntity entity =
        associatedDeviceDatabase.getAssociatedDeviceChallengeSecret(deviceId);
    if (entity == null) {
      logd(TAG, "Challenge secret not found!");
      return null;
    }

    return cryptoHelper.decrypt(entity.encryptedChallengeSecret);
  }

  /**
   * Hash provided value with device's challenge secret and return result. Returns {@code null} if
   * unsuccessful.
   */
  @Nullable
  public byte[] hashWithChallengeSecret(@NonNull String deviceId, @NonNull byte[] value) {
    byte[] challengeSecret = getChallengeSecret(deviceId);
    if (challengeSecret == null) {
      loge(TAG, "Unable to find challenge secret for device " + deviceId + ".");
      return null;
    }

    Mac mac;
    try {
      mac = Mac.getInstance(CHALLENGE_HASHING_ALGORITHM);
    } catch (NoSuchAlgorithmException e) {
      loge(TAG, "Unable to find hashing algorithm " + CHALLENGE_HASHING_ALGORITHM + ".", e);
      return null;
    }

    SecretKeySpec keySpec = new SecretKeySpec(challengeSecret, CHALLENGE_HASHING_ALGORITHM);
    try {
      mac.init(keySpec);
    } catch (InvalidKeyException e) {
      loge(TAG, "Exception while initializing HMAC.", e);
      return null;
    }

    return mac.doFinal(value);
  }

  @NonNull
  private SharedPreferences getSharedPrefs() {
    // This should be called only after user 0 is unlocked.
    if (sharedPreferences != null) {
      return sharedPreferences;
    }
    sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE);
    return sharedPreferences;
  }

  /**
   * Get the unique id for head unit. Persists on device until factory reset. This should be called
   * only after user 0 is unlocked.
   *
   * @return unique id
   */
  @NonNull
  public UUID getUniqueId() {
    if (uniqueId != null) {
      return uniqueId;
    }

    SharedPreferences prefs = getSharedPrefs();
    if (prefs.contains(UNIQUE_ID_KEY)) {
      uniqueId = UUID.fromString(prefs.getString(UNIQUE_ID_KEY, null));
      logd(TAG, "Found existing trusted unique id: " + prefs.getString(UNIQUE_ID_KEY, ""));
    }

    if (uniqueId == null) {
      uniqueId = UUID.randomUUID();
      prefs.edit().putString(UNIQUE_ID_KEY, uniqueId.toString()).apply();
      logd(TAG, "Generated new trusted unique id: " + prefs.getString(UNIQUE_ID_KEY, ""));
    }

    return uniqueId;
  }

  /**
   * Get a list of associated devices for the given user.
   *
   * @param userId The identifier of the user.
   * @return Associated device list.
   */
  @NonNull
  public List<AssociatedDevice> getAssociatedDevicesForUser(@NonNull int userId) {
    List<AssociatedDeviceEntity> entities =
        associatedDeviceDatabase.getAssociatedDevicesForUser(userId);

    if (entities == null) {
      return new ArrayList<>();
    }

    ArrayList<AssociatedDevice> userDevices = new ArrayList<>();
    for (AssociatedDeviceEntity entity : entities) {
      userDevices.add(entity.toAssociatedDevice());
    }

    return userDevices;
  }

  /**
   * Get a list of associated devices for the current user.
   *
   * @return Associated device list.
   */
  @NonNull
  public List<AssociatedDevice> getActiveUserAssociatedDevices() {
    return getAssociatedDevicesForUser(ActivityManager.getCurrentUser());
  }

  /**
   * Returns a list of device ids of associated devices for the given user.
   *
   * @param userId The user id for whom we want to know the device ids.
   * @return List of device ids.
   */
  @NonNull
  public List<String> getAssociatedDeviceIdsForUser(@NonNull int userId) {
    List<AssociatedDevice> userDevices = getAssociatedDevicesForUser(userId);
    ArrayList<String> userDeviceIds = new ArrayList<>();

    for (AssociatedDevice device : userDevices) {
      userDeviceIds.add(device.getDeviceId());
    }

    return userDeviceIds;
  }

  /**
   * Retrieves devices associated with the active user.
   *
   * @param listener {@link OnAssociatedDevicesRetrievedListener} that will be notified when devices
   *     are retrieved.
   */
  public void retrieveForActiveUserAssociatedDevice(OnAssociatedDevicesRetrievedListener listener) {
    listener.onAssociatedDevicesRetrieved(getActiveUserAssociatedDevices());
  }

  /**
   * Returns a list of device ids of associated devices for the current user.
   *
   * @return List of device ids.
   */
  @NonNull
  public List<String> getActiveUserAssociatedDeviceIds() {
    return getAssociatedDeviceIdsForUser(ActivityManager.getCurrentUser());
  }

  /**
   * Add the associated device of the given deviceId for the currently active user.
   *
   * @param device New associated device to be added.
   */
  public void addAssociatedDeviceForActiveUser(@NonNull AssociatedDevice device) {
    addAssociatedDeviceForUser(ActivityManager.getCurrentUser(), device);
    if (associatedDeviceCallback != null) {
      associatedDeviceCallback.onAssociatedDeviceAdded(device);
    }
  }

  /**
   * Add the associated device of the given deviceId for the given user.
   *
   * @param userId The identifier of the user.
   * @param device New associated device to be added.
   */
  public void addAssociatedDeviceForUser(int userId, @NonNull AssociatedDevice device) {
    AssociatedDeviceEntity entity =
        new AssociatedDeviceEntity(userId, device, /* isConnectionEnabled= */ true);
    associatedDeviceDatabase.addOrReplaceAssociatedDevice(entity);
  }

  /**
   * Update the name for an associated device.
   *
   * @param deviceId The id of the associated device.
   * @param name The name to replace with. Empty names are ignored.
   */
  public void updateAssociatedDeviceName(@NonNull String deviceId, @NonNull String name) {
    if (name.isEmpty()) {
      logw(TAG, "Attempted to update the device name to an empty string. Ignoring.");
      return;
    }
    AssociatedDeviceEntity entity = associatedDeviceDatabase.getAssociatedDevice(deviceId);
    if (entity == null) {
      logw(TAG, "Attempted to update name on an unrecognized device " + deviceId + ". Ignoring.");
      return;
    }
    updateName(entity, name);
  }

  /**
   * Set the name for an associated device only if it does not already have a name populated.
   *
   * @param deviceId The id of the associated device.
   * @param name The name to set on the associated device. Empty names are ignored.
   */
  public void setAssociatedDeviceName(@NonNull String deviceId, @NonNull String name) {
    if (name.isEmpty()) {
      logw(TAG, "Attempted to set the device name to an empty string. Ignoring.");
      return;
    }
    AssociatedDeviceEntity entity = associatedDeviceDatabase.getAssociatedDevice(deviceId);
    if (entity == null) {
      logw(TAG, "Attempted to set name on an unrecognized device " + deviceId + ". Ignoring.");
      return;
    }
    if (entity.name != null) {
      logd(TAG, "Name was already set for device " + deviceId + ". No further action taken.");
      return;
    }
    updateName(entity, name);
  }

  private void updateName(AssociatedDeviceEntity entity, String name) {
    entity.name = name;
    associatedDeviceDatabase.addOrReplaceAssociatedDevice(entity);
    if (associatedDeviceCallback != null) {
      associatedDeviceCallback.onAssociatedDeviceUpdated(
          new AssociatedDevice(entity.id, entity.address, name, entity.isConnectionEnabled));
    }
  }

  /**
   * Remove the associated device of the given deviceId for the given user.
   *
   * @param userId The identifier of the user.
   * @param deviceId The identifier of the device to be cleared.
   */
  public void removeAssociatedDevice(int userId, @NonNull String deviceId) {
    AssociatedDeviceEntity entity = associatedDeviceDatabase.getAssociatedDevice(deviceId);
    if (entity == null || entity.userId != userId) {
      return;
    }
    associatedDeviceDatabase.removeAssociatedDevice(entity);
    if (associatedDeviceCallback != null) {
      associatedDeviceCallback.onAssociatedDeviceRemoved(
          new AssociatedDevice(deviceId, entity.address, entity.name, entity.isConnectionEnabled));
    }
  }

  /**
   * Clear the associated device of the given deviceId for the current user.
   *
   * @param deviceId The identifier of the device to be cleared.
   */
  public void removeAssociatedDeviceForActiveUser(@NonNull String deviceId) {
    removeAssociatedDevice(ActivityManager.getCurrentUser(), deviceId);
  }

  /**
   * Set if connection is enabled for an associated device.
   *
   * @param deviceId The id of the associated device.
   * @param isConnectionEnabled If connection enabled for this device.
   */
  public void updateAssociatedDeviceConnectionEnabled(
      @NonNull String deviceId, boolean isConnectionEnabled) {
    AssociatedDeviceEntity entity = associatedDeviceDatabase.getAssociatedDevice(deviceId);
    if (entity == null) {
      logw(
          TAG,
          "Attempt to enable or disable connection on an unrecognized device "
              + deviceId
              + ". Ignoring.");
      return;
    }
    if (entity.isConnectionEnabled == isConnectionEnabled) {
      return;
    }
    entity.isConnectionEnabled = isConnectionEnabled;
    associatedDeviceDatabase.addOrReplaceAssociatedDevice(entity);
    if (associatedDeviceCallback != null) {
      associatedDeviceCallback.onAssociatedDeviceUpdated(
          new AssociatedDevice(deviceId, entity.address, entity.name, isConnectionEnabled));
    }
  }

  /**
   * Get associated device with the given id.
   *
   * @param deviceId The id of the associated device.
   * @return Associated device.
   */
  @Nullable
  public AssociatedDevice getAssociatedDevice(@NonNull String deviceId) {
    AssociatedDeviceEntity entity = associatedDeviceDatabase.getAssociatedDevice(deviceId);
    if (entity == null) {
      logw(TAG, "No device has been associated with device id " + deviceId + ". Returning null");
      return null;
    }
    return new AssociatedDevice(deviceId, entity.address, entity.name, entity.isConnectionEnabled);
  }

  /** Callback for association device related events. */
  public interface AssociatedDeviceCallback {
    /** Triggered when an associated device has been added. */
    void onAssociatedDeviceAdded(@NonNull AssociatedDevice device);

    /** Triggered when an associated device has been removed. */
    void onAssociatedDeviceRemoved(@NonNull AssociatedDevice device);

    /** Triggered when an associated device has been updated. */
    void onAssociatedDeviceUpdated(@NonNull AssociatedDevice device);
  }

  /** Listener for retrieving devices associated with the active user. */
  public interface OnAssociatedDevicesRetrievedListener {

    /** Triggered when the devices associated with the active user are retrieved. */
    void onAssociatedDevicesRetrieved(List<AssociatedDevice> devices);
  }
}

File: libs/connecteddevice/src/com/google/android/connecteddevice/storage/ConnectedDeviceStorage.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.connecteddevice.storage

import android.app.ActivityManager
import android.content.Context
import android.content.SharedPreferences
import androidx.room.Room
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.google.android.companionprotos.DeviceOS
import com.google.android.connecteddevice.model.AssociatedDevice
import com.google.android.connecteddevice.util.SafeLog.logd
import com.google.android.connecteddevice.util.SafeLog.loge
import com.google.android.connecteddevice.util.SafeLog.logw
import com.google.android.connecteddevice.util.ThreadSafeCallbacks
import java.security.InvalidKeyException
import java.security.InvalidParameterException
import java.security.NoSuchAlgorithmException
import java.util.UUID
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Storage for connected devices in a car. */
open class ConnectedDeviceStorage(
  private val context: Context,
  private val cryptoHelper: CryptoHelper,
  private val associatedDeviceDatabase: AssociatedDeviceDao,
  private val callbackExecutor: Executor,
) {
  /** Callback for association device related events. */
  interface AssociatedDeviceCallback {
    fun onAssociatedDeviceAdded(device: AssociatedDevice)

    fun onAssociatedDeviceRemoved(device: AssociatedDevice)

    fun onAssociatedDeviceUpdated(device: AssociatedDevice)
  }

  /** Listener for retrieving devices associated with the active user. */
  interface OnAssociatedDevicesRetrievedListener {
    /** Triggered when the devices associated with the active user are retrieved. */
    fun onAssociatedDevicesRetrieved(devices: List<AssociatedDevice>)
  }

  private val sharedPreferences: SharedPreferences by lazy {
    // This should be called only after user 0 is unlocked.
    context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE)
  }
  private val callbacks = ThreadSafeCallbacks<AssociatedDeviceCallback>()

  constructor(
    context: Context
  ) : this(
    context,
    KeyStoreCryptoHelper(),
    Room.databaseBuilder(context, ConnectedDeviceDatabase::class.java, DATABASE_NAME)
      .addMigrations(MIGRATION_2_3)
      .fallbackToDestructiveMigration(dropAllTables = true)
      .build()
      .associatedDeviceDao(),
    Executors.newSingleThreadExecutor(),
  )

  open val uniqueId: UUID by lazy {
    var uuid: UUID? = sharedPreferences.getString(UNIQUE_ID_KEY, null)?.let { UUID.fromString(it) }

    if (uuid == null) {
      uuid = UUID.randomUUID()
      sharedPreferences.edit().putString(UNIQUE_ID_KEY, uuid.toString()).apply()
      logd(TAG, "Generated new trusted unique id: $uuid")
    }

    uuid
  }

  /** Registers an [AssociatedDeviceCallback] for associated device updates. */
  open fun registerAssociatedDeviceCallback(callback: AssociatedDeviceCallback) {
    callbacks.add(callback, callbackExecutor)
  }

  /** Unregisters an [AssociatedDeviceCallback] from associated device updates. */
  open fun unregisterAssociatedDeviceCallback(callback: AssociatedDeviceCallback) {
    callbacks.remove(callback)
  }

  /** Returns the encryption key for [deviceId]; `null` if not recognized. */
  open suspend fun getEncryptionKey(deviceId: String): ByteArray? {
    val entity = associatedDeviceDatabase.getAssociatedDeviceKey(deviceId)
    if (entity == null) {
      logd(TAG, "Encryption key not found!")
      return null
    }
    return cryptoHelper.decrypt(entity.encryptedKey)
  }

  /** Saves the encryption key for the given deviceId. */
  open suspend fun saveEncryptionKey(deviceId: String, encryptionKey: ByteArray) {
    val encryptedKey = cryptoHelper.encrypt(encryptionKey)
    val entity = AssociatedDeviceKeyEntity(deviceId, encryptedKey)

    associatedDeviceDatabase.addOrReplaceAssociatedDeviceKey(entity)
    logd(TAG, "Successfully wrote encryption key for $deviceId.")
  }

  /**
   * Saves the challenge secret for the given deviceId.
   *
   * @param secret Secret associated with this device. Note: must be [CHALLENGE_SECRET_BYTES] bytes
   *   in length or an [InvalidParameterException] will be thrown.
   */
  open suspend fun saveChallengeSecret(deviceId: String, secret: ByteArray) {
    if (secret.size != CHALLENGE_SECRET_BYTES) {
      throw InvalidParameterException("Secrets must be $CHALLENGE_SECRET_BYTES bytes in length.")
    }
    val encryptedKey = cryptoHelper.encrypt(secret)
    val entity = AssociatedDeviceChallengeSecretEntity(deviceId, encryptedKey)

    associatedDeviceDatabase.addOrReplaceAssociatedDeviceChallengeSecret(entity)
    logd(TAG, "Successfully wrote challenge secret for $deviceId.")
  }

  /** Returns the challenge secret associated with the deviceId; `null` if not recognized. */
  suspend fun getChallengeSecret(deviceId: String): ByteArray? {
    val entity = associatedDeviceDatabase.getAssociatedDeviceChallengeSecret(deviceId)
    if (entity == null) {
      logd(TAG, "Challenge secret not found!")
      return null
    }

    return cryptoHelper.decrypt(entity.encryptedChallengeSecret)
  }

  /**
   * Hashes the [value] with device's challenge secret and returns result. Returns [null] if
   * unsuccessful.
   */
  open suspend fun hashWithChallengeSecret(deviceId: String, value: ByteArray): ByteArray? {
    val challengeSecret = getChallengeSecret(deviceId)
    if (challengeSecret == null) {
      loge(TAG, "Unable to find challenge secret for device $deviceId.")
      return null
    }

    val mac =
      try {
        Mac.getInstance(CHALLENGE_HASHING_ALGORITHM)
      } catch (e: NoSuchAlgorithmException) {
        loge(TAG, "Unable to find hashing algorithm $CHALLENGE_HASHING_ALGORITHM", e)
        return null
      }

    val keySpec = SecretKeySpec(challengeSecret, CHALLENGE_HASHING_ALGORITHM)
    try {
      mac.init(keySpec)
    } catch (e: InvalidKeyException) {
      loge(TAG, "Exception while initializing HMAC.", e)
      return null
    }

    return mac.doFinal(value)
  }

  /** Gets a list of all the associated devices. */
  open suspend fun getAllAssociatedDevices(): List<AssociatedDevice> {
    val entities = associatedDeviceDatabase.getAllAssociatedDevices() ?: emptyList()

    return entities.map { it.toAssociatedDevice() }
  }

  /** Gets a list of associated devices for the given user. */
  suspend fun getAssociatedDevicesForUser(userId: Int): List<AssociatedDevice> {
    val entities = associatedDeviceDatabase.getAssociatedDevicesForUser(userId) ?: emptyList()

    return entities.map { it.toAssociatedDevice() }
  }

  /** Gets a list of associated devices for the current driver. */
  open suspend fun getDriverAssociatedDevices(): List<AssociatedDevice> {
    return getAssociatedDevicesForUser(ActivityManager.getCurrentUser())
  }

  /** Gets a list of associated devices for all passengers. */
  open suspend fun getPassengerAssociatedDevices(): List<AssociatedDevice> {
    return getAssociatedDevicesNotBelongingToUser(ActivityManager.getCurrentUser())
  }

  // Used by test.
  internal suspend fun getAssociatedDevicesNotBelongingToUser(userId: Int): List<AssociatedDevice> {
    val entities = associatedDeviceDatabase.getAllAssociatedDevices() ?: emptyList()

    return entities.filter { it.userId != userId }.map { it.toAssociatedDevice() }
  }

  /** Returns a list of device ids of associated devices for the given userId. */
  open suspend fun getAssociatedDeviceIdsForUser(userId: Int): List<String> {
    val userDevices = getAssociatedDevicesForUser(userId)

    return userDevices.map { it.id }
  }

  /** Returns a list of device ids of associated devices for the current driver. */
  suspend fun getDriverAssociatedDeviceIds(): List<String> {
    return getAssociatedDeviceIdsForUser(ActivityManager.getCurrentUser())
  }

  /** Returns a list of device ids of associated devices for all passengers. */
  suspend fun getPassengerAssociatedDeviceIds(): List<String> {
    return getAssociatedDeviceIdsNotBelongingToUser(ActivityManager.getCurrentUser())
  }

  internal suspend fun getAssociatedDeviceIdsNotBelongingToUser(userId: Int): List<String> {
    val entities = associatedDeviceDatabase.getAllAssociatedDevices() ?: emptyList()

    return entities.filter { it.userId != userId }.map { it.id }
  }

  /** Adds the associated device of the given deviceId for the current driver. */
  open suspend fun addAssociatedDeviceForDriver(device: AssociatedDevice) {
    addAssociatedDeviceForUser(ActivityManager.getCurrentUser(), device)
  }

  /** Adds the associated device for the given user. */
  open suspend fun addAssociatedDeviceForUser(userId: Int, device: AssociatedDevice) {
    val entity = AssociatedDeviceEntity(userId, device, /* isConnectionEnabled= */ true)
    addOrReplaceAssociatedDevice(entity)
    callbacks.invoke { it.onAssociatedDeviceAdded(device) }
  }

  /** Updates the name for an associated device. */
  open suspend fun updateAssociatedDeviceName(deviceId: String, name: String) {
    if (name.isEmpty()) {
      logw(TAG, "Attempted to update the device name to an empty string. Ignoring.")
      return
    }
    updateAssociatedDevice(deviceId) { it.name = name }
  }

  /**
   * Sets the name for an associated device only if it does not already have a name populated.
   *
   * Empty names are ignored.
   */
  suspend fun setAssociatedDeviceName(deviceId: String, name: String) {
    if (name.isEmpty()) {
      logw(TAG, "Attempted to set the device name to an empty string. Ignoring.")
      return
    }
    updateAssociatedDevice(deviceId) { entity ->
      if (entity.name != null) {
        logd(TAG, "Name was already set for device $deviceId. No further action taken.")
        return@updateAssociatedDevice
      }
      entity.name = name
    }
  }

  open suspend fun updateAssociatedDeviceOs(deviceId: String, deviceOs: DeviceOS) {
    updateAssociatedDevice(deviceId) { it.os = deviceOs }
  }

  open suspend fun updateAssociatedDeviceOsVersion(deviceId: String, deviceOsVersion: String) {
    updateAssociatedDevice(deviceId) { it.osVersion = deviceOsVersion }
  }

  open suspend fun updateAssociatedDeviceCompanionSdkVersion(deviceId: String, sdkVersion: String) {
    updateAssociatedDevice(deviceId) { it.companionSdkVersion = sdkVersion }
  }

  /** Removes the associated device of the given deviceId. */
  open suspend fun removeAssociatedDevice(deviceId: String) {
    val entity = associatedDeviceDatabase.getAssociatedDevice(deviceId)
    if (entity == null) {
      return
    }
    associatedDeviceDatabase.removeAssociatedDevice(entity)
    callbacks.invoke { it.onAssociatedDeviceRemoved(entity.toAssociatedDevice()) }
  }

  /** Sets if connection is enabled for an associated device. */
  open suspend fun updateAssociatedDeviceConnectionEnabled(
    deviceId: String,
    isConnectionEnabled: Boolean,
  ) {
    updateAssociatedDevice(deviceId) { it.isConnectionEnabled = isConnectionEnabled }
  }

  /** Returns the associated device with the given deviceId. */
  open suspend fun getAssociatedDevice(deviceId: String): AssociatedDevice? {
    val entity = associatedDeviceDatabase.getAssociatedDevice(deviceId)
    if (entity == null) {
      logw(TAG, "No device has been associated with device id $deviceId. Returning null.")
      return null
    }
    return entity.toAssociatedDevice()
  }

  /** Updates the identified associated device to be claimed by the current user. */
  open suspend fun claimAssociatedDevice(deviceId: String) {
    logd(TAG, "Claiming device $deviceId for the current user.")
    updateAssociatedDevice(deviceId) { it.userId = ActivityManager.getCurrentUser() }
  }

  /** Removes the claim on the identified associated device leaving it in an unclaimed state. */
  open suspend fun removeAssociatedDeviceClaim(deviceId: String) {
    logd(TAG, "Removing the user claim for device $deviceId.")
    updateAssociatedDevice(deviceId) { it.userId = AssociatedDevice.UNCLAIMED_USER_ID }
  }

  private suspend fun updateAssociatedDevice(
    deviceId: String,
    update: (entity: AssociatedDeviceEntity) -> Unit,
  ) {
    val entity = associatedDeviceDatabase.getAssociatedDevice(deviceId)
    if (entity == null) {
      logw(TAG, "Could not retrieve device with $deviceId. Ignoring.")
      return
    }

    update(entity)

    addOrReplaceAssociatedDevice(entity)
    callbacks.invoke { it.onAssociatedDeviceUpdated(entity.toAssociatedDevice()) }
  }

  private suspend fun addOrReplaceAssociatedDevice(entity: AssociatedDeviceEntity) {
    // Needed for database 2_3 migration.
    if (entity.os == null) {
      entity.os = DeviceOS.DEVICE_OS_UNKNOWN
    }
    associatedDeviceDatabase.addOrReplaceAssociatedDevice(entity)
  }

  companion object {
    private const val TAG = "CompanionStorage"

    private const val SHARED_PREFS_NAME = "com.google.android.connecteddevice"
    private const val UNIQUE_ID_KEY = "CTABM_unique_id"
    private const val DATABASE_NAME = "connected-device-database"

    private const val CHALLENGE_HASHING_ALGORITHM = "HmacSHA256"

    const val CHALLENGE_SECRET_BYTES = 32

    // Database migration from version 2 to 3.
    // This migration adds the os, osVersion, and companionSdkVersion columns to the
    // associated_devices table.
    private val MIGRATION_2_3 =
      object : Migration(2, 3) {
        override fun migrate(database: SupportSQLiteDatabase) {
          database.execSQL(
            "ALTER TABLE associated_devices ADD os TEXT NOT NULL DEFAULT 'DEVICE_OS_UNKNOWN';"
          )
          database.execSQL("ALTER TABLE associated_devices ADD osVersion TEXT;")
          database.execSQL("ALTER TABLE associated_devices ADD companionSdkVersion TEXT;")
        }
      }
  }
}
